src/arfsdao.ts

```diff
@@ -750,74 +810,391 @@ export class ArFSPrivateDrive extends ArFSEntity implements ArFSDriveEntity {
+	}
+
+	private static setupNodesWithEntity(
+		entity: ArFSFileOrFolderEntity,
+		folderIdToEntityMap: { [k: string]: ArFSFileOrFolderEntity },
+		folderIdToNodeMap: { [k: string]: FolderTreeNode }
+	): void {
+		const folderIdKeyIsPresent = Object.keys(folderIdToNodeMap).includes(entity.entityId);
+		const parentFolderIdKeyIsPresent = Object.keys(folderIdToNodeMap).includes(entity.parentFolderId);
+		if (!folderIdKeyIsPresent) {
+			if (!parentFolderIdKeyIsPresent) {
+				const parentFolderEntity = folderIdToEntityMap[entity.parentFolderId];
+				if (parentFolderEntity) {
+					this.setupNodesWithEntity(parentFolderEntity, folderIdToEntityMap, folderIdToNodeMap);
+				}
 			}
+			const parent = folderIdToNodeMap[entity.parentFolderId];
+			if (parent) {
+				const node = new FolderTreeNode(entity.entityId, parent);
+				parent.children.push(node);
+				folderIdToNodeMap[entity.entityId] = node;
+			} else {
+				// this one is supposed to be the new root
+				const rootNode = new FolderTreeNode(entity.entityId);
+				folderIdToNodeMap[entity.entityId] = rootNode;
+			}
+		}
+	}
 
-			return new ArFSPrivateDrive(
-				this.appName,
-				this.appVersion,
-				this.arFS,
-				this.contentType,
-				this.driveId,
-				this.entityType,
-				this.name,
-				this.txId,
-				this.unixTime,
-				this.drivePrivacy,
-				this.rootFolderId,
-				this.driveAuthMode,
-				this.cipher,
-				this.cipherIV
-			);
+	public get rootNode(): FolderTreeNode {
+		if (this._rootNode) {
+			return this._rootNode;
+		}
+
+		const someFolderId = Object.keys(this.folderIdToEntityMap)[0];
+		let tmpNode = this.folderIdToNodeMap[someFolderId];
+		while (tmpNode.parent && this.folderIdToNodeMap[tmpNode.parent.folderId]) {
+			tmpNode = tmpNode.parent;
+		}
+		this._rootNode = tmpNode;
+		return tmpNode;
+	}
+
+	public subTreeOf(folderId: FolderID, maxDepth = Number.POSITIVE_INFINITY): FolderHierarchy {
+		const newRootNode = this.folderIdToNodeMap[folderId];
+
+		const subTreeNodes = this.nodeAndChildrenOf(newRootNode, maxDepth);
+
+		const entitiesMapping = subTreeNodes.reduce((accumulator, node) => {
+			return Object.assign(accumulator, { [node.folderId]: this.folderIdToEntityMap[node.folderId] });
+		}, {});
+		const nodesMapping = subTreeNodes.reduce((accumulator, node) => {
+			return Object.assign(accumulator, { [node.folderId]: node });
+		}, {});
+
+		return new FolderHierarchy(entitiesMapping, nodesMapping);
+	}
+
+	public allFolderIDs(): FolderID[] {
+		return Object.keys(this.folderIdToEntityMap);
+	}
+
+	public nodeAndChildrenOf(node: FolderTreeNode, maxDepth: number): FolderTreeNode[] {
+		const subTreeEntities: FolderTreeNode[] = [node];
+		if (maxDepth > 0) {
+			node.children.forEach((child) => {
+				subTreeEntities.push(...this.nodeAndChildrenOf(child, maxDepth - 1));
+			});
+		}
+		return subTreeEntities;
+	}
+
+	public subFolderIDsForFolderId(folderId: FolderID, maxDepth: number): FolderID[] {
```

Comment:
While reading through some of the code I think this might be better if named `folderIdSubtreeFromFolderId`

Reply:
I agree. a115d7bde95d2ce03b7cc8f57e6dc8890d1ca5e7